src/test/java/io/axoniq/axonserver/connector/DoOnFirstResultStreamTest.java

```diff
@@ -0,0 +1,158 @@
+        when(delegate.nextIfAvailable(anyLong(), any(TimeUnit.class))).thenReturn(ELEMENT);
+        when(delegate.next()).thenReturn(ELEMENT);
+        when(delegate.isClosed()).thenReturn(CLOSED);
+        when(delegate.getError()).thenReturn(Optional.empty());
+        //noinspection unchecked
+        doOnFirstCallback = mock(Consumer.class);
+        testSubject = new DoOnFirstResultStream<>(delegate, doOnFirstCallback);
+    }
+
+    @Test
+    void testOnFirstCallbackInvokedWhenPeeked() {
+        assertEquals(ELEMENT, testSubject.peek());
+        verify(doOnFirstCallback).accept(ELEMENT);
+
+        assertEquals(ELEMENT, testSubject.peek());
+        verifyNoMoreInteractions(doOnFirstCallback);
+        verify(delegate, times(2)).peek();
+    }
+
+    @Test
+    void testOnFirstCallbackInvokedWhenNextIfAvailable() {
+        assertEquals(ELEMENT, testSubject.nextIfAvailable());
+        verify(doOnFirstCallback).accept(ELEMENT);
+
+        assertEquals(ELEMENT, testSubject.nextIfAvailable());
+        verifyNoMoreInteractions(doOnFirstCallback);
+        verify(delegate, times(2)).nextIfAvailable();
+    }
+
+    @Test
+    void testOnFirstCallbackInvokedWhenNextIfAvailableWithTimeout() throws InterruptedException {
+        assertEquals(ELEMENT, testSubject.nextIfAvailable(1, TimeUnit.SECONDS));
+        verify(doOnFirstCallback).accept(ELEMENT);
+
+        assertEquals(ELEMENT, testSubject.nextIfAvailable(1, TimeUnit.SECONDS));
+        verifyNoMoreInteractions(doOnFirstCallback);
+        verify(delegate, times(2)).nextIfAvailable(1, TimeUnit.SECONDS);
+    }
+
+    @Test
+    void testOnFirstCallbackInvokedWhenNext() throws InterruptedException {
+        assertEquals(ELEMENT, testSubject.next());
+        verify(doOnFirstCallback).accept(ELEMENT);
+
+        assertEquals(ELEMENT, testSubject.next());
+        verifyNoMoreInteractions(doOnFirstCallback);
+        verify(delegate, times(2)).next();
+    }
+
+    @Test
+    void testDelegationOfNonConsumableMethods() {
+        assertTrue(testSubject.isClosed());
+        assertEquals(Optional.empty(), testSubject.getError());
+        testSubject.close();
+        testSubject.onAvailable(null);
+
+        verify(delegate).isClosed();
+        verify(delegate).getError();
+        verify(delegate).close();
+        verify(delegate).onAvailable(null);
+    }
+
+    @RepeatedTest(1000)
+    void testOnFirstIsCalledOnlyOnce() throws InterruptedException {
+        ExecutorService executorService = Executors.newFixedThreadPool(4);
+        List<Runnable> actions = asList(() -> testSubject.peek(),
+                                        this::silentNext,
+                                        () -> testSubject.nextIfAvailable(),
+                                        this::silentNextIfAvailable);
+        Collections.shuffle(actions);
+
+        actions.stream()
+               .map(executorService::submit)
+               .forEach(this::completeFuture);
```

Comment:
Minor this, but the `completeFuture` name is misleading. The method doesn't "complete" the future, it awaits for it to be completed. I suggest: `awaitComplete`.